### PR TITLE
dynamo: wrap graph break inst in try except block - with context manager setup/teardown

### PR DESCRIPTION
**Pull request by jon-chuang**
Replacement to https://github.com/pytorch/pytorch/pull/94672. 

Follow up to https://github.com/pytorch/pytorch/pull/94137.

We simply replace the set grad mode try except blocks with one for a more generic contextmanager (using `__enter__` and `__exit__`), storing the context manager into a `symbolic_local` for the duration of the try block.

(see https://github.com/pytorch/torchdynamo/issues/207 for the original motivation)

This allows us to handle calling inner functions with graph breaks for any arbitrarily deep nesting of live context managers subclassing `AbstractContextManager`. (see tests)

cc @soumith @voznesenskym @yanboliang @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @desertfire @jansel @ngimel @Chillee

### CODE REVIEW

File: torch/_dynamo/resume_execution.py [hunk 1]
```diff
@@ -28,8 +28,81 @@
 @dataclasses.dataclass(frozen=True)
 class ReenterWith:
     stack_index: int = None
+    target_values: Optional[Tuple] = None
+
+    # If we do not want to destroy the stack, we can do the same thing as a
+    # `SETUP_WITH` block, only that we store the context manager in a local_symbol
+    def try_except(self, stack_len, code_options, cleanup: List[Instruction]):
+        load_args = []
```
Comment (jon-chuang) [inline torch/_dynamo/resume_execution.py:36]: Handle python <= 3.8 as well, following prev example

File: torch/_dynamo/resume_execution.py [hunk 2]
```diff
@@ -133,7 +208,7 @@ def update(instructions: List[Instruction], code_options: Dict[str, Any]):
             freevars = tuple(code_options["co_cellvars"] or []) + tuple(
                 code_options["co_freevars"] or []
             )
-            code_options["co_name"] = f"<graph break in {code_options['co_name']}>"
+            code_options["co_name"] = f"<resume in {code_options['co_name']}>"
```
Comment (jon-chuang) [inline torch/_dynamo/resume_execution.py:211]: calling the resume function graph break is misleading. The code stub that calls resume is what deserves to be called graph break.

File: torch/_dynamo/symbolic_convert.py [hunk 3]
```diff
@@ -1007,11 +998,12 @@ def CALL_FUNCTION_KW(self, inst):
         self.call_function(fn, args, kwargs)
 
     def LOAD_METHOD(self, inst):
+        obj = self.stack[-1].clone()
```
Comment (jon-chuang) [inline torch/_dynamo/symbolic_convert.py:1001]: Fixed an unused code path

File: torch/_dynamo/symbolic_convert.py [hunk 4]
```diff
@@ -1761,11 +1753,7 @@ def inline_call(cls, parent, func, args, kwargs):
             return cls.inline_call_(parent, func, args, kwargs)
 
     @staticmethod
-    def inline_call_(parent, func, args, kwargs):
-        assert isinstance(
-            func,
-            (UserFunctionVariable, NestedUserFunctionVariable),
-        )
+    def check_inlineable(func):
```
Comment (jon-chuang) [inline torch/_dynamo/symbolic_convert.py:1756]: Just some tidying up

File: torch/_dynamo/bytecode_analysis.py [hunk 5]
```diff
@@ -175,7 +175,6 @@ def stacksize_analysis(instructions):
     if False:
         for inst in instructions:
             stack_size = stack_sizes[inst]
-            print(stack_size.low, stack_size.high, inst)
```
Comment (jon-chuang) [inline torch/_dynamo/bytecode_analysis.py:178]: remove a stray print. Should this be made a log.debug?

File: torch/_dynamo/symbolic_convert.py [hunk 6]
```diff
@@ -1007,11 +998,13 @@ def CALL_FUNCTION_KW(self, inst):
         self.call_function(fn, args, kwargs)
 
     def LOAD_METHOD(self, inst):
+        unimplemented("LOAD_METHOD should have been converted to LOAD_ATTR")
```
Comment (jon-chuang) [inline torch/_dynamo/symbolic_convert.py:1001]: Explanation for review: Fixed an unused code path

File: torch/_dynamo/resume_execution.py [hunk 7]
```diff
@@ -185,13 +268,11 @@ def update(instructions: List[Instruction], code_options: Dict[str, Any]):
...
     def unreachable_codes(code_options):
         """Codegen a `raise None` to make analysis work for unreachable code"""
-        if None not in code_options["co_consts"]:
-            code_options["co_consts"] = tuple(code_options["co_consts"]) + (None,)
         return [
             create_instruction(
                 "LOAD_CONST",
                 argval=None,
-                arg=code_options["co_consts"].index(None),
+                arg=PyCodegen.get_const_index(code_options, None),
```
Comment (jon-chuang) [inline torch/_dynamo/resume_execution.py:275]: Explanation for review: fix lack of usage of existing API

**Comment by yanboliang** [comment]
@jon-chuang This PR mixed the changes from #94137, can you remove what has been included in #94137? I'll review after that. Thank you!

**Comment by jon-chuang** [comment]
@yanboliang, it depends on that PR, so let's wait for that one to be merged, I'll rebase.

File: torch/_dynamo/resume_execution.py [hunk 8]
```diff
@@ -31,6 +31,76 @@ class ReenterWith:
...
+        ]
+        if sys.version_info < (3, 9):
+            epilogue = [
+                create_instruction("POP_BLOCK"),
+                PyCodegen.create_begin_finally(),
+                except_jump_target,
+                *reset,
+                create_instruction("END_FINALLY"),
+            ]
+        else:
```
Comment (jon-chuang) [inline torch/_dynamo/resume_execution.py:90]: Let's add 3.11 support here.

Comment (jon-chuang) [reply torch/_dynamo/resume_execution.py:90]: Seems to have been unnecessary. Cool

File: torch/_dynamo/bytecode_analysis.py (hunk 5 above)
Comment (williamwen42) [reply torch/_dynamo/bytecode_analysis.py:179]: The entire block `if False:` doesn't run, leading me to believe it's debug code. It's been helpful for me when debugging, so I would hesitate on removing it.

File: torch/_dynamo/bytecode_transformation.py [hunk 9]
```diff
@@ -397,7 +397,12 @@ def fix_vars(instructions: List[Instruction], code_options):
         if instructions[i].opcode in HAS_LOCAL:
             instructions[i].arg = varnames[instructions[i].argval]
         elif instructions[i].opcode in HAS_NAME:
-            instructions[i].arg = names[instructions[i].argval]
+            if not instructions[i].argval in names:
```
Comment (williamwen42) [inline torch/_dynamo/bytecode_transformation.py:400]: Other places in the code add names to `code_options["co_names"]`. We should either assert for existence here; or only add global names here and not elsewhere.

File: torch/_dynamo/codegen.py [hunk 10]
```diff
@@ -343,5 +343,6 @@ def load_import_from(self, module_name, object_name):
             )
         )
 
-    def create_begin_finally(self):
+    @staticmethod
+    def create_begin_finally():
```
Comment (williamwen42) [inline torch/_dynamo/codegen.py:347]: Can we delete this function?

File: torch/_dynamo/resume_execution.py (hunk 8 above)
Comment (williamwen42) [reply torch/_dynamo/resume_execution.py:90]: I have a bunch of 3.11 changes affecting this part that I'm currently trying to merge.

File: torch/_dynamo/symbolic_convert.py (hunk 6 above)
Comment (williamwen42) [reply torch/_dynamo/symbolic_convert.py:1005]: This is used in 3.11.

File: torch/_dynamo/bytecode_analysis.py (hunk 5 above)
Comment (jansel) [reply torch/_dynamo/bytecode_analysis.py:179]: Let's either leave this in, or remove the entire `if False:` block.  The entire loop becomes pointless without the print.

Comment (jon-chuang) [reply torch/_dynamo/bytecode_analysis.py:179]: I'll clean it up the formatting and put the predicate in the config so it's more visible. (Another PR so we don't clobber)

File: torch/_dynamo/resume_execution.py [hunk 11]
```diff
@@ -31,6 +31,76 @@ class ReenterWith:
...
+        if self.target_values:
+            load_args = [
+                create_instruction(
+                    "LOAD_CONST",
+                    PyCodegen.get_const_index(code_options, val),
+                    val,
+                )
+                for val in self.target_values
+            ]
+        ctx_name = f"___context_manager_{self.stack_index}"
```
Comment (yanboliang) [inline torch/_dynamo/resume_execution.py:47]: This can be duplicated, can you use this helper function?
https://github.com/pytorch/pytorch/blob/34a7c79eac1b0743c76043b81cd2b32326cb9035/torch/_dynamo/bytecode_transformation.py#L580-L581

**Comment by yanboliang** [comment]
Gently ping @jon-chuang, can you address these inline comments? We hit another bug #94758, which potentially can be fixed by this. It's good we can merge this ASAP.

File: torch/_dynamo/bytecode_transformation.py (hunk 9 above)
Comment (jon-chuang) [reply torch/_dynamo/bytecode_transformation.py:482]: I think this may be an unnecessary artifact from a earlier separate attempt

Comment (jon-chuang) [reply torch/_dynamo/bytecode_transformation.py:482]: reverted and added to code_options on per-needed basis.

**Comment by jon-chuang** [comment]
Hello @yanboliang , should be ready now.

**Comment by jon-chuang** [comment]
We need to add a `not-user-facing` label

File: test/dynamo/test_misc.py [hunk 12]
```diff
@@ -3374,6 +3396,109 @@ def fn(x):
...
+
+        module = MyModule()
+        real_16, real_32 = module(torch.tensor([0.5]))
+        real_device_16 = real_16.device
+        real_dtype_16 = real_16.dtype
+        real_device_32 = real_32.device
+        real_dtype_32 = real_32.dtype
+
+        graph = torch._dynamo.optimize("eager")(module)
+        exported_16, exported_32 = graph(torch.tensor([0.5]))
```
Comment (yanboliang) [inline test/dynamo/test_misc.py:3435]: You are not using ```torch._dynamo.export```, so it's better to name it as ```expected_16``` rather than ```exported_16``` to reduce confusion.

**Review by yanboliang — APPROVED** [review]
LGTM, cc @williamwen42 to check if this breaks any Python 3.11 change.

**Comment by jon-chuang** [comment]
> if this breaks any Python 3.11 change.

I guess full 3.11 testing is not currently in CI? If it is, it seems safe to merge.

**Comment by jon-chuang** [comment]
> Else if it is, it seems safe to merge.

Anw, if it turns out there is a problem, we can fix it, I assume CI is the source of truth here.
